designs/0035-pre-model-gqs.md

```diff
@@ -0,0 +1,556 @@
+through the latent data block.  Nevertheless, computation in the
+latent data block can affect what happens in the model block by
+defining new variables on which the model block may condition.
+
+### Representing the latent data
+
+The design is based on an object-based representation of the latent
+data using code generation.  For example, suppose the latent data
+block is defined as follows.
+
+
+```cpp
+latent data {
+  real<lower=0, upper=1> sens = beta_rng(98, 2);
+  real<lower=0, upper=1> spec = beta_rng(95, 5);
+}
+```
+
+This will lead to the code generation of a class in the generated
+`.hpp` file for the model.
+
+```cpp
+struct LatentData {
+  int iteration_number__;  // included by default in all LatentData
+  double sens;
+  double spec;
+};
+```
+
+### Generating the latent data
+
+An additional method in the model class will be required to populate 
+the latent data. 
+
+```cpp
+template <typename RNG> inline void
+latent_data(
+    const VectorXd& params_r,
+    LatentData& latent_data__,
+    RNG& base_rng,
+    int iteration = -1,
+    std::ostream* pstream=nullptr) const {
+  latent_data__.iteration_number__ = iteration;
+  ... code generated to populate latent_data__ ...
+}
+```
+
+The context will be responsible for managing the memory of any data  
+structures in LatentData, such as `Eigen::Matrix` or `std::vector`
+instances. By reusing the same `LatentData` variable (per thread),
+variables can be reset rather than allocating and freeing each evaluation.
+
+
+### Modified log_prob method
+
+The existing `log_prob` method must be updated to maintain backward
+compatibility: 
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_
+log_prob(Eigen::Matrix<T_,-1,1>& params_r, 
+         const LatentData& latent_data__,
+         std::ostream* pstream = nullptr) const;
+```
+
+In this case, the `log_prob` code generation must be updated to copy
+all of the generated data into scope.
+
+```cpp
+template <bool propto__, bool jacobian__, typename T_> inline T_ 
+log_prob(
+    Eigen::Matrix<T_,-1,1>& params_r,  
+    const LatentData& latent_data  , 
+    std::ostream* pstream = nullptr) const {
+  double sens = latent_data__.sens;
+  double spec = latent_data__.spec;
+  ... code generate as before with sens/spec in scope ...
+}
+```
+
+For ease of backward compatibility, we can have the old implementation
+with a default value,
+
+```cpp  
+template <bool propto__, bool jacobian__, typename T_> inline T_ 
+log_prob(
+    Eigen::Matrix<T_,-1,1>& params_r,  
+    std::ostream* pstream = nullptr) const {
+  LatentData latent_data,
+  return log_prob(params_r, latent_data, pstream) const;
+```
+
+### BridgeStan
+
+BridgeStan can be updated to mimic the updates to the Stan model
+class.
+
+# Drawbacks
+[drawbacks]: #drawbacks
+
+1. It will require additional documentation and tests, increasing
+   Stan's long-term maintenance burden and the overhead required for
+   someone to understand what it's doing. 
```

Comment:
This drawback is too generic to be a useful discussion point. It should not be the first item in the list, if included at all.

Reply:
I think it's worth keeping, so I'll move it down.  I just always want to remind people of that.  Too many people think it's just the initial development cost for an algorithm (I have this discussion *all the time* with people).

Reply:
P.S.  Thanks for all the super useful comments, @nhuure!